sallyport/src/lib.rs

```diff
@@ -205,18 +207,19 @@ impl Cursor<'_> {
             padding.len() + size_of::<T>() * count
         };
 
-        let (data, next) = (*inner_ptr).split_at_mut(mid);
+        let (data, next) = self.0.split_at_mut(mid);
 
-        *inner_ptr = next;
-
-        Ok(data.align_to_mut::<T>().1)
+        Ok((Cursor(next), data.align_to_mut::<T>().1))
     }
 
     /// Copies data from a value into a slice using self.alloc().
-    pub fn copy_slice<T: Copy>(&self, value: &[T]) -> core::result::Result<&mut [T], ()> {
-        let slice: &mut [T] = unsafe { self.alloc(value.len())? };
+    pub fn copy_slice<T: Copy>(
+        self,
+        value: &[T],
+    ) -> core::result::Result<(Cursor<'a>, &'a mut [T]), ()> {
```

Comment:
Explicit elision?

Reply:
Hmm, I'm not sure you can apply explicit elision here since we do have two lifetimes going on here: 1) is the lifetime bound to cursor *and* returned slice, and 2) is the lifetime bound to the input slice. I think we are fine here saying the two do not necessarily have to be equal, but rather, one can be a subset of the other, right? In this case, explicit elision will collapse both lifetimes into one and the same.

Reply:
Which goes to my overall concern posted at the top of the review. I'm concerned this patch might allow the allocated slice to live longer than the cursor.

Reply:
Oh, the output slice is still bound by the same lifetime as the original block stored in the first `Cursor`. Effectively what @haraldh has done here was adapting the [`std::slice::split_at_mut`](https://doc.rust-lang.org/stable/std/primitive.slice.html#method.split_at_mut) pattern in `Cursor`. What this implies is that we cannot hold multiple mutable borrows to overlapping slices (i.e., alias the memory), but we can to non-overlapping chunks.